src/main/java/org/jenkinsci/plugins/workflow/support/storage/BulkFlowNodeStorage.java

```diff
@@ -0,0 +1,247 @@
+            flush();
+        }
+    }
+
+    @Override
+    public void storeNode(FlowNode n) throws IOException {
+        storeNode(n, false);
+    }
+
+    /**
+     * Persists a single FlowNode to disk (if not already persisted).
+     * @param n Node to persist
+     * @throws IOException
+     */
+    @Override
+    public void flushNode(@Nonnull FlowNode n) throws IOException {
+        storeNode(n, false);
+    }
+
+    /** Force persisting any nodes that had writing deferred */
+    @Override
+    public void flush() throws IOException {
+        if (nodes != null && isModified) {
+            // TODO reuse a single buffer if we can, and consider using async FileChannel operations.
+            if (!dir.exists()) {
+                IOUtils.mkdirs(dir);
+            }
+            PipelineIOUtils.writeByXStream(nodes, getStoreFile(), XSTREAM, !this.isAvoidAtomicWrite());
+
+            isModified = false;
+        }
+    }
+
+    public List<Action> loadActions(@Nonnull FlowNode node) throws IOException {
+        Tag t = getOrLoadNodes().get(node.getId());
+        return (t != null) ? t.actions() : Collections.<Action>emptyList();
+    }
+
+    /**
+     * Just stores this one node
+     */
+    public void saveActions(@Nonnull FlowNode node, @Nonnull List<Action> actions) throws IOException {
+        HashMap<String, Tag> map = getOrLoadNodes();
+        Tag t = map.get(node.getId());
+        if (t != null) {
+            t.node = node;
+            List<Action> act = node.getActions();
+            t.actions = act.toArray(new Action[act.size()]);
+        } else {
+            map.put(node.getId(), new Tag(node, actions));
+        }
+        isModified = true;
+    }
+
+    /** Have we written everything to disk that we need to, or is there something waiting to be written */
+    public boolean isPersistedFully() {
+        return !isModified;
+    }
+
+
+    /**
+     * To group node and their actions together into one object.
+     */
+    private static class Tag {
+        /* @Nonnull except perhaps after deserialization */ FlowNode node;
+        private @CheckForNull Action[] actions;
+
+        private Tag(@Nonnull FlowNode node, @Nonnull List<Action> actions) {
+            this.node = node;
+            this.actions = actions.isEmpty() ? null : actions.toArray(new Action[actions.size()]);
+        }
+
+        private void storeActions() {  // We've already loaded the actions, may as well store them to the FlowNode
+            try {
+                FlowNode_setActions.invoke(this.node, actions());
+            } catch (InvocationTargetException|IllegalAccessException ex) {
+                throw new RuntimeException(ex);
+            }
+        }
+
+        public @Nonnull List<Action> actions() {
+            return actions != null ? Arrays.asList(actions) : Collections.<Action>emptyList();
+        }
+    }
+
+    public static final XStream2 XSTREAM = new XStream2();
+
+    private static final Field FlowNode$exec;
```

Comment:
Ah, ignore my earlier dollar sign question. That said, I’d prefer if we didn’t use uncommon characters and formatting in variable names.